rasterio/rio/features.py

```diff
@@ -20,6 +19,128 @@
 logger = logging.getLogger('rio')
 
 
+# Extract command
+@cli.command(short_help='Extract raster using features.')
+@files_inout_arg
+@format_opt
+@click.option('--all_touched', is_flag=True, default=False,
+              help='Use all pixels touched by features for extraction, '
+              'otherwise use only pixels whose center is within the '
+              'polygon or that are selected by Brezenhams line algorithm')
+@click.option('--crop', is_flag=True, default=False,
+              help='Crop output raster to the extent of the geometries. '
+                   'GeoJSON must overlap input raster to use --crop')
+@click.option('--invert', is_flag=True, default=False,
+              help='Inverts the mask, so that areas covered by features are'
+                   'masked out and areas not covered are retained.  Ignored '
+                   'if using --crop')
+@click.pass_context
+def extract(
+        ctx,
+        files,
+        driver,
+        all_touched,
+        crop,
+        invert):
+
+    """Extracts pixels from all bands of a raster using features, masking out
+    all areas not covered by features, and optionally crops the output raster
+    to the extent of the features.  Features are assumed to be in the same
+    coordinate reference system as the input raster.
+
+    GeoJSON must be the first input file or provided from stdin:
+
+    > rio extract features.json input.tif output.tif
+
+    > rio extract input.tif output.tif < features.json
+
+    If the output raster exists, it will be completely overwritten with the
+    results of this operation.
+
+    The result is always equal to or within the bounds of the input raster.
+
+    --crop and --invert options are mutually exclusive.
+
+    --crop option is not valid if features are completely outside extent of
+    input raster.
+    """
```

Comment:
How about `-i/--invert` and `-j/--geojson-file`?

Yes, I see the difference in between rio-rasterize and what I'm suggesting. Piping in GeoJSON would be nice, I agree. I'm concerned it interferes with being able to pipe or pipeline rasters (#188), which is what we should be aiming for in the suite of rio commands. In the rio-rasterize case, there aren't any raster inputs, so it's fine to turn everything over to GeoJSON and have it be the foremost input.

I'm also a tiny bit concerned about confusion around the term "extract". We already refer to feature extraction in a bunch of places (like the rio-shapes docs). Could this be better called "rio-select" or "rio-mask"?